File: module-app/src/main/java/maple/expectation/scheduler/ExpectationCalculationScheduler.java
package maple.expectation.scheduler;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.infrastructure.executor.LogicExecutor;
import maple.expectation.infrastructure.executor.TaskContext;
import maple.expectation.infrastructure.persistence.repository.GameCharacterRepository;
import maple.expectation.service.v5.queue.PriorityCalculationQueue;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

/**
 * V5 CQRS: Expectation Calculation Scheduler
 *
 * <p>Periodically adds LOW priority tasks to refresh all users' expectation calculations.
 *
 * <h3>Flow</h3>
 *
 * <ol>
 *   <li>Fetch all user IGNs from database (paginated)
 *   <li>Add LOW priority task to queue for each user
 *   <li>Worker processes tasks in background
 * </ol>
 *
 * <h3>Scheduling Strategy</h3>
 *
 * <ul>
 *   <li>fixedDelay: Waits specified time AFTER previous execution completes
 *   <li>Prevents queue overload by respecting backpressure
 *   <li>Batch size configurable via application.yml
 * </ul>
 *
 * <h3>Use Cases</h3>
 *
 * <ul>
 *   <li>Nightly full recalculation of all users
 *   <li>Periodic data freshness maintenance
 *   <li>Background cache warming
 * </ul>
 */
@Slf4j
@Component
@RequiredArgsConstructor
@ConditionalOnProperty(
    name = "scheduler.expectation-calculation.enabled",
    havingValue = "true",
    matchIfMissing = false)
public class ExpectationCalculationScheduler {

  private final PriorityCalculationQueue queue;
  private final GameCharacterRepository gameCharacterRepository;
  private final LogicExecutor executor;

  /**
   * Batch size for expectation calculation scheduler.
   *
   * <p>P2 Fix: Use @Value injection instead of parsing placeholder string. Previous code:
   * Integer.parseInt("${scheduler.expectation-calculation.batch-size:100}")
   */
  @org.springframework.beans.factory.annotation.Value(
      "${scheduler.expectation-calculation.batch-size:100}")
  private int batchSize;

  /**
   * Add LOW priority tasks for all users
   *
   * <p>Configuration:
   *
   * <ul>
   *   <li>scheduler.expectation-calculation.fixed-delay-ms: Delay between executions (default:
   *       3600000 = 1 hour)
   *   <li>scheduler.expectation-calculation.batch-size: Users per batch (default: 100)
   * </ul>
   *
   * <h3>Backpressure Handling</h3>
   *
   * <p>If queue is full (MAX_QUEUE_SIZE = 10,000), LOW priority tasks are rejected gracefully
   * without error.
   */
  @Scheduled(fixedDelayString = "${scheduler.expectation-calculation.fixed-delay-ms:3600000}")
  public void refreshAllUsers() {
    TaskContext context = TaskContext.of("Scheduler", "ExpectationCalculation.RefreshAll");

    executor.executeVoid(
        () -> {
          log.info("[ExpectationCalculation] Starting full user refresh");
          int processedCount = 0;
          int skippedCount = 0;

          // Paginated fetch to avoid memory issues
          final int pageSize = 100;
          int page = 0;
          boolean hasMore = true;

          while (hasMore) {
            final int currentPage = page;
            hasMore =
                executor.executeOrDefault(
                    () -> {
                      org.springframework.data.domain.PageRequest pageRequest =
                          org.springframework.data.domain.PageRequest.of(currentPage, pageSize);
                      return gameCharacterRepository.findAll(pageRequest).stream()
                          .map(character -> addTaskForUser(character.getUserIgn()))
                          .toList()
                          .contains(true);
                    },
                    false,
                    context);

            if (hasMore) {
              page++;
              processedCount += pageSize;

              // Log progress every batch size users
              if (processedCount % batchSize == 0) {
                log.info(
                    "[ExpectationCalculation] Processed {} users, skipped {} (queue full)",
                    processedCount,
                    skippedCount);
              }
            }
          }

          log.info(
              "[ExpectationCalculation] Full refresh completed: processed={}, skipped={}, queueSize={}",
              processedCount,
              skippedCount,
              queue.size());
        },
        context);
  }

  /**
   * Add LOW priority task for a single user
   *
   * @return true if task was added, false if rejected (backpressure)
   */
  private boolean addTaskForUser(String userIgn) {
    return queue.addLowPriorityTask(userIgn);
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/scheduler/ExpectationCalculationScheduler.kt
package maple.expectation.infrastructure.scheduler

import maple.expectation.core.domain.model.Page
import maple.expectation.core.domain.model.PageRequest
import maple.expectation.core.domain.model.character.GameCharacter
import maple.expectation.core.port.out.QueueWriterPort
import maple.expectation.infrastructure.executor.LogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import maple.expectation.infrastructure.persistence.repository.GameCharacterRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

/**
 * V5 CQRS: Expectation Calculation Scheduler (ADR-005 이관)
 */
@Component
@ConditionalOnProperty(
    name = ["scheduler.expectation-calculation.enabled"],
    havingValue = "true",
    matchIfMissing = false,
)
class ExpectationCalculationScheduler(
    private val queueWriter: QueueWriterPort,
    private val gameCharacterRepository: GameCharacterRepository,
    private val executor: LogicExecutor,
) {
    private val log = LoggerFactory.getLogger(ExpectationCalculationScheduler::class.java)

    @Value("\${scheduler.expectation-calculation.batch-size:100}")
    private var batchSize: Int = 100

    @Scheduled(fixedDelayString = "\${scheduler.expectation-calculation.fixed-delay-ms:3600000}")
    fun refreshAllUsers() {
        val context = TaskContext.of("Scheduler", "ExpectationCalculation.RefreshAll")

        executor.executeVoidJava(
            {
                log.info("[ExpectationCalculation] Starting full user refresh")
                var processedCount = 0
                var skippedCount = 0

                val pageSize = 100
                var page = 0
                var hasMore = true

                while (hasMore) {
                    val currentPage = page
                    hasMore = executor.executeOrDefault(
                        {
                            val pageRequest: PageRequest = PageRequest.of(currentPage, pageSize)
                            val characterPage: Page<GameCharacter> =
                                gameCharacterRepository.findAll(pageRequest)

                            for (character in characterPage.content) {
                                addTaskForUser(character.userIgn.value)
                            }

                            characterPage.hasNext
                        },
                        false,
                        context,
                    )

                    if (hasMore) {
                        page++
                        processedCount += pageSize

                        if (processedCount % batchSize == 0) {
                            log.info(
                                "[ExpectationCalculation] Processed {} users, skipped {} (queue full)",
                                processedCount,
                                skippedCount,
                            )
                        }
                    }
                }

                log.info(
                    "[ExpectationCalculation] Full refresh completed: processed={}, skipped={}, queueSize={}",
                    processedCount,
                    skippedCount,
                    queueWriter.size(),
                )
            },
            context,
        )
    }

    private fun addTaskForUser(userIgn: String): Boolean = queueWriter.addLowPriorityTask(userIgn)
}
